### PR TITLE
feature / lz gated modules

### PR DESCRIPTION
**Pull request by imthatcarlos**
## New features
- new modules `LZGatedFollowModule`, `LZGatedCollectModule`, and `LZGatedReferenceModule` to do cross-chain token  gating
- supporting lz contracts `LZGatedProxy`, `LzApp`, `NonblockingLzApp`, and `SimpleLzApp`
- new mocha tests
- new hardhat tasks

## Example (`LZGatedFollowModule`)
lz-gated-follow-module@2x (1)
1. **lensprotocol.lens** has set their gated follow data. A new follower must have 50 AAVE on Optimism.
2. **carlosbeltran.lens** wishes to follow - fetches the requirements and generates the signature for `#followWithSig`
3. anyone (ex: a relayer) can submit the follow requirements + the signature to the `LZGatedProxy` contract on Optimism, which relays the payload to the `LZGatedFollowModule` contract on Polygon via LayerZero
- TLDR; the token balance check happens on Optimism, and the follow tx on Polygon

## TODO
- [ ] fix newly failing test in `/lz-gated-follow-module.spec.ts`
- [x] fix cause for event `MessageFailed` being emitted when the [tx actually succeeds](https://mumbai.polygonscan.com/tx/0x9954807dd4493610d4fdb9b5309e20dbdf2d358aeef66f94039317187ba01f39)
- [x] improve fee estimates
- [x] improve gas estimates
- [x] ask lz team: latency between `#send` on source chain and `#lzReceive` on destination chain can be between 5-30min+ (goerli => mumbai) => testnet infra was updated, not an issue on mainnet ✅

### CODE REVIEW

File: contracts/lz/LzApp.sol [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import "./interfaces/ILayerZeroReceiver.sol";
+import "./interfaces/ILayerZeroUserApplicationConfig.sol";
+import "./interfaces/ILayerZeroEndpoint.sol";
+
+/**
+ * @title LzApp
+ * @notice LayerZero-enabled contract that can have multiple remote chain ids.
+ * @dev this is a modified contract from the layerzero suggested implementation
+ */
+abstract contract LzApp is Owned, ILayerZeroReceiver, ILayerZeroUserApplicationConfig {
```
Comment (donosonaumczuk) [inline contracts/lz/LzApp.sol]: Please, instead of creating a new file, let's import `LzApp` adding LayerZero repo as dependency

File: contracts/lz/NonblockingLzApp.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import "./LzApp.sol";
+
+// https://github.com/LayerZero-Labs/solidity-examples/blob/main/contracts/lzApp/NonblockingLzApp.sol
+
+/*
+ * the default LayerZero messaging behaviour is blocking, i.e. any failed message will block the channel
+ * this abstract class try-catch all fail messages and store locally for future retry. hence, non-blocking
+ * NOTE: if the srcAddress is not configured properly, it will still block the message pathway from (srcChainId, srcAddress)
+ */
+abstract contract NonblockingLzApp is LzApp {
```
Comment (donosonaumczuk) [inline contracts/lz/NonblockingLzApp.sol]: Please, instead of creating a new file, let's import `NonblockingLzApp` adding LayerZero repo as dependency

File: contracts/lz/SimpleLzApp.sol [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import "./interfaces/ILayerZeroUserApplicationConfig.sol";
+import "./interfaces/ILayerZeroEndpoint.sol";
+
+/**
+ * @title SimpleLzApp
+ * @notice LayerZero-enabled contract that has only one trusted remote chain and sends messages cross-chain (does NOT
+ * receive) to remote contracts provided by the concrete class.
+ * NOTE: this contract is ownable only to set layerzero configs.
+ */
+abstract contract SimpleLzApp is Owned, ILayerZeroUserApplicationConfig {
```
Comment (donosonaumczuk) [inline contracts/lz/SimpleLzApp.sol]: Please, instead of creating a new file, let's import `SimpleLzApp` adding LayerZero repo as dependency

File: contracts/lz/interfaces/ILayerZeroEndpoint.sol [hunk 4]
```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.5.0;
+
+import "./ILayerZeroUserApplicationConfig.sol";
+
+interface ILayerZeroEndpoint is ILayerZeroUserApplicationConfig {
```
Comment (donosonaumczuk) [inline contracts/lz/interfaces/ILayerZeroEndpoint.sol]: Please, instead of creating a new file, let's import `ILayerZeroEndpoint` adding LayerZero repo as dependency

File: contracts/lz/interfaces/ILayerZeroReceiver.sol [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.5.0;
+
+interface ILayerZeroReceiver {
```
Comment (donosonaumczuk) [inline contracts/lz/interfaces/ILayerZeroReceiver.sol]: Please, instead of creating a new file, let's import `ILayerZeroReceiver` adding LayerZero repo as dependency

File: contracts/lz/interfaces/ILayerZeroUserApplicationConfig.sol [hunk 6]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.5.0;
+
+interface ILayerZeroUserApplicationConfig {
```
Comment (donosonaumczuk) [inline contracts/lz/interfaces/ILayerZeroUserApplicationConfig.sol]: Please, instead of creating a new file, let's import `ILayerZeroUserApplicationConfig` adding LayerZero repo as dependency

File: contracts/lz/libraries/LzLib.sol [hunk 7]
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity >=0.6.0;
+pragma experimental ABIEncoderV2;
+
+library LzLib {
```
Comment (donosonaumczuk) [inline contracts/lz/libraries/LzLib.sol]: Please let's also import this from a dependency instead of adding the file

File: contracts/lz/mocks/LZEndpointMock.sol [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+like a real LayerZero endpoint but can be mocked, which handle message transmission, verification, and receipt.
+- blocking: LayerZero provides ordered delivery of messages from a given sender to a destination chain.
+- non-reentrancy: endpoint has a non-reentrancy guard for both the send() and receive(), respectively.
+- adapter parameters: allows UAs to add arbitrary transaction params in the send() function, like airdrop on destination chain.
+unlike a real LayerZero endpoint, it is
+- no messaging library versioning
+- send() will short circuit to lzReceive()
+- no user application configuration
+*/
+contract LZEndpointMock is ILayerZeroEndpoint {
```
Comment (donosonaumczuk) [inline contracts/lz/mocks/LZEndpointMock.sol]: Please let's also import this from a dependency instead of adding the file

File: contracts/mocks/ERC20Mock.sol [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract ERC20Mock is ERC20, Ownable {
```
Comment (donosonaumczuk) [inline contracts/mocks/ERC20Mock.sol]: We already have an ERC20 mock. Let's avoid adding this duplicated mock please
https://github.com/lens-protocol/modules/blob/master/contracts/mocks/ACurrency.sol

File: contracts/mocks/ERC721Mock.sol [hunk 10]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+contract ERC721Mock is ERC721, Ownable {
```
Comment (donosonaumczuk) [inline contracts/mocks/ERC721Mock.sol]: We also have a mock for ERC721s, I think it suits the use case too. So let's use the already existent instead, please
https://github.com/lens-protocol/modules/blob/master/contracts/mocks/NFT.sol

**Review by donosonaumczuk — CHANGES_REQUESTED** [review]
Hey guys, good work!

There are lot of files that we are adding to the repo that are not necessarily ad-hoc for the Lens modules, but instead generic LayerZero contracts. It would be nice if you import your core repo as a dependency instead, or even create a separate repo that can be used as dependency if some of the files are not included in your core repo. 

The idea is to avoid increasing the amount of files here when are not 100% related to the code of each of the Lens modules. Same reason why I am asking to remove some duplicated mocks from the PR.

**Comment by imthatcarlos** [comment]
thanks for the feedback @donosonaumczuk 🙏 I resolved all comments and pushed changes 
- imported the `@layerzerolabs/solidity-examples` package for the lz contracts + deleted all the extra files
- opted to re-using the `LzApp` instead of my `SimpleLzApp`; all i had to do was not process messages received
- used the exiting mocks
- add `LzEndpointMock` in the typechain config externalArtifacts; only runs for tests, lmk if that works
